readthedocs/projects/views/public.py

```diff
@@ -106,45 +107,53 @@ def get_context_data(self, **kwargs):
         return context
 
 
-def _badge_return(redirect, url):
-    if redirect:
-        return HttpResponseRedirect(url)
+def _badge_return(version, status):
+    template = loader.get_template('badges/badge.svg')
+    if status == 'passing':
+        fill = '#4c1'
+    elif status == 'failing':
+        fill = '#e05d44'
     else:
-        response = requests.get(url)
-        http_response = HttpResponse(response.content,
-                                     content_type="image/svg+xml")
-        http_response['Cache-Control'] = 'no-cache'
-        http_response['Etag'] = md5.new(url)
-        return http_response
+        fill = '#dfb317'
+    content = template.render({
+        'project': version.project.slug,
+        'project_length': len(version.project.slug) * 5,
+        'project_fill': len(version.project.slug) * 5 / 2 + 1,
+        'version': version.slug,
+        'version_length': len(version.slug) * 5,
+        'version_fill': len(version.slug) * 5 / 2 + 1,
+        'status': status,
+        'fill': fill,
+        'full_width': (len(version.project.slug) + len(version.slug)) * 5,
+    })
+    http_response = HttpResponse(content,
+                                 content_type="image/svg+xml")
+    http_response['Cache-Control'] = 'no-cache'
+    http_response['Etag'] = md5.new(content)
```

Comment:
`md5` module has been deprecated since Python 2.5. I'd replace it with `hashlib.md5(content).hexdigest()`.